rosetta-source/src/main/rosetta/event-common-func.rosetta

```diff
@@ -1828,6 +1828,451 @@ func Qualify_Shaping: <"The qualification of a shaping event from the fact that
+		newCollateralPortfolio CollateralPortfolio(1..1) <"New collateral portfolio to subtitute for the original collateral.">
+		priceQuantity PriceQuantity (1..*) <"The price and quantity of the substituted product.">
+	output: 
+		instruction PrimitiveInstruction (1..1)
+
+	condition ContractualProduct: <"Only a repo contractual product can substitute collateral.">
+		tradeState -> trade -> tradableProduct -> product -> contractualProduct exists
+
+        set instruction: <"Sets the second part of the split to be a new contract with a new asset payout.">
+                PrimitiveInstruction {
+                    quantityChange: QuantityChangeInstruction {
+                            change: priceQuantity,
+                            direction: QuantityChangeDirectionEnum -> Replace,
+                            lotIdentifier: empty
+                        },
+                    termsChange: Create_SubstitutionInstruction(
+                                tradeState -> trade -> tradableProduct -> product -> contractualProduct,
+                               effectiveDate,
+                               newCollateralPortfolio
+                            ),
+                    ...
+                }
+
+func Create_SubstitutionInstruction:<"Creates the terms change instruction that updates the payout with the new subsitution payout.">
+	inputs:
+		contractualProduct ContractualProduct (1..1) <"The orginal contractual product to be used as the basis of the new trade.">
+		effectiveDate AdjustableOrRelativeDate (1..1) <"The date to close and open a new position.">
+		newCollateralPortfolio CollateralPortfolio(1..1) <"New collateral portfolio to subtitute for the original collateral.">
+	output:
+		termsChangeInstruction TermsChangeInstruction (1..1)
+
+	set termsChangeInstruction -> product -> contractualProduct:
+		contractualProduct
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> effectiveDate:
+		effectiveDate
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> collateral -> collateralPortfolio:
+		newCollateralPortfolio
+
+
+func Qualify_Substitution: <"Qualification of a collateral substitution event.">
+	[qualification BusinessEvent]
+	inputs:
+	businessEvent BusinessEvent (1..1)
+	output: is_event boolean (1..1)
+
+	alias beforeEconomicterms: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias openEconomicTerms: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias closedTradeState: FilterClosedTradeStates( businessEvent -> after )
+    
+	set is_event:
+		beforeEconomicterms exists and		
+		openEconomicTerms exists and
+		closedTradeState count = 1 and
+		openEconomicTerms -> payout <> beforeEconomicterms -> payout and
```

Comment:
`payout` should be identical before and after, only `collateral` changes

Reply:
The type of collateral is substituted but the amount and price would be different.

Reply:
That's shouldn't affect the `payout` attribute, if the amount and price only concern `collateral`

Reply:
The payout is modified because the term is shorter than the original.

Reply:
done